api/jobset/v1alpha2/jobset_types.go

```diff
@@ -51,6 +52,11 @@ const (
 	CoordinatorKey = "jobset.sigs.k8s.io/coordinator"
 )
 
+var (
+	// the legacy names are no longer defined in the api, only in k/2/apis/batch
```

Comment:
Can we not introduce the legacy labels now? I think we can remove the original ones now that 1.27 is out of support.

Reply:
I think it is better to copy them, because Job controller still sets them. 

Here is an example Pod template created on cluster 1.30.2:
```
        batch.kubernetes.io/controller-uid: c46c3b0d-76d8-427b-bda6-a9cb06c77cfd
        batch.kubernetes.io/job-name: sample-job-a-lv6wd
        controller-uid: c46c3b0d-76d8-427b-bda6-a9cb06c77cfd
        job-name: sample-job-a-lv6wd
```
If they are not present on the original Job template (null) I wouldn't copy them to the newJob template.

Reply:
Moreover we still expect them in the Job validation: https://github.com/kubernetes/kubernetes/blob/60c4c2b2521fb454ce69dee737e3eb91a25e0535/pkg/apis/batch/validation/validation.go#L139-L146

Reply:
Hm, but if I don't copy them here, then we would update Job template not to contain them, I also not sure this will pass validation, let me test.

Reply:
I opened up https://github.com/kubernetes-sigs/jobset/pull/642 to drop JobNameKey as it wasn't actually used at all in JobSet.

Reply:
I think maybe we should just copy all labels and not just the managed ones.

If a user has user specified labels, why would we drop those in this case?

Reply:
Ah I see. I missed that this is resume.

Reply:
```
		// Copy the Job labels managed by the Job controller
		for _, labelKey := range jobset.JobManagedLabels {
			newJob.Spec.Template.Labels[labelKey] = job.Spec.Template.Labels[labelKey]
		}
		job.Spec.Template.Labels = collections.MergeMaps(
			newJob.Spec.Template.Labels,
			rjob.Template.Spec.Template.Labels,
		)
```

I think that we are already copying all labels. I guess my point is why don't we just copy all labels from job.Spec.Templates.Labels[labelKey]?

We should not need to expose the details of those labels and copy all labels.

Reply:
> If a user has user specified labels, why would we drop those in this case?

This seems like a bug to me. A user creates a JobSet and sets labels in their template. I don't expect that they should ever be dropped?

Reply:
The code you quote here:
```
		job.Spec.Template.Labels = collections.MergeMaps(
			newJob.Spec.Template.Labels,
			rjob.Template.Spec.Template.Labels,
		)
```
merges labels from `newJob` and `rjob` (Job template in JobSet). None of the two places contain the labels in question. The labels are only present on the old Job, and so I copy from them.

Reply:
So could we copy all the labels from the old Job without looping over the special managedByJob ones?

Reply:
> We should not need to expose the details of those labels and copy all labels.

For now I don't see another way of fixing the bug, either:
1. copy over the Job labels, 
2. delete the Jobs and recreate the Jobs, letting API server default all the fields / labels

This is partly why I prefer (2.) - and implemented initially. However, I understand that updating the Jobs might be preferred as deleting and recreating is a bigger change.

Reply:
My point is that 1 could be done just by copying ALL labels from the previous job.

Reply:
> So could we copy all the labels from the old Job without looping over the special managedByJob ones?

Consider the following transitions by Kueue: resume (RF1) -> suspend -> resume (RF2).

It is possible that resuming on RF2 does not add the same labels as for RF1. If we copied all labels, then we would always trail the old ones from RF1.

Indeed, this is mostly problem for `nodeSelectors`, where Kueue may want to choose a new set of nodeSelectors for RF2. I would be ok to accept this issue for labels as it does not have much of a practical issue for Kueue currently, but it would be a "known issue", and may hit us one day.

Reply:
> My point is that 1 could be done just by copying ALL labels from the previous job.

Seems like a race condition :). Consider the scenario I describe here: https://github.com/kubernetes-sigs/jobset/pull/640#discussion_r1711840934.  The code will then remain problematic if RF1 adds a label, which is not expected in RF2, because with this approach we would never get rid of it. 

It will remain problematic in some corner cases and hard to debug. It will also be very surprising why we restore nodeSelectors, but not labels.

Reply:
I synced with @kannon92 on slack and there is a simpler solution which I implement here: https://github.com/kubernetes-sigs/jobset/pull/644. It does not cover fully the case of "restoring" a PodTemplate, but it should cover most cases where the admission to RF2 overrides the previous values.

This is needed anyway as the first step, and part of this PR. We can return to these bits if really shown they are needed.